proxy/src/task.rs

```diff
@@ -0,0 +1,212 @@
+//! Task execution utilities.
+use futures::future::{
+    Future,
+    ExecuteError,
+    ExecuteErrorKind,
+    Executor,
+};
+use tokio::{
+    executor::{
+        DefaultExecutor,
+        Executor as TokioExecutor,
+        SpawnError,
+    },
+    runtime::{self as thread_pool, current_thread},
+};
+
+use std::error::Error;
+use std::{fmt, io};
+
+/// An empty type which implements `Executor` by lazily  calling
+/// `tokio::executor::DefaultExecutor::current().execute(...)`.
+///
+/// This can be used when we would simply like to call `tokio::spawn` rather
+/// than explicitly using a particular executor, but need an `Executor` for
+/// a generic type or to pass to a function which expects one.
+///
+/// Note that this uses `DefaultExecutor` rather than `tokio::spawn`, as we
+/// would prefer for our `Executor` implementation to pass errors rather than
+/// panicking (as `tokio::spawn` does).
+#[derive(Copy, Clone, Debug, Default)]
+pub struct LazyExecutor;
+
+/// Indicates which Tokio `Runtime` should be used for the main proxy.
+///
+/// This is either a `tokio::runtime::current_thread::Runtime`, or a
+/// `tokio::runtime::Runtime` (thread pool). This type simply allows
+/// both runtimes to present a unified interface, so that they can be
+/// used to construct a `Main`.
+///
+/// This allows the runtime used for the proxy to be customized based
+/// on the application: for a sidecar proxy, we use the current thread
+/// runtime, but for an ingress proxy, we would prefer the thread pool.
+pub enum MainRuntime {
+    CurrentThread(current_thread::Runtime),
+    ThreadPool(thread_pool::Runtime),
+}
+
+/// Like a `SpawnError` or `ExecuteError`, but with an implementation
+/// of `std::error::Error`.
+#[derive(Debug, Clone)]
+pub enum TaskError {
```

Comment:
tioli, I think it would be fine to just name this `Error` and it can be referred to as `task::Error`

Reply:
Yeah, wanted to import `std::error::Error` here w/o renaming, but I think that's a good call.